apps/heft/src/operations/OperationExecutionManager.ts

```diff
@@ -0,0 +1,235 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as os from 'os';
+import { performance } from 'perf_hooks';
+import { AlreadyReportedError, Async, ITerminal } from '@rushstack/node-core-library';
+
+import { AsyncOperationQueue, IOperationSortFunction } from './AsyncOperationQueue';
+import { OperationStatus } from './OperationStatus';
+import { IOperationExecutionRecordContext, OperationExecutionRecord } from './OperationExecutionRecord';
+import type { Operation } from './Operation';
+import type { LoggingManager } from '../pluginFramework/logging/LoggingManager';
+
+export interface IOperationExecutionManagerOptions {
+  debugMode: boolean;
+  parallelism: string | undefined;
+  terminal: ITerminal;
+  loggingManager: LoggingManager;
+}
+
+/**
+ * A class which manages the execution of a set of tasks with interdependencies.
+ * Initially, and at the end of each task execution, all unblocked tasks
+ * are added to a ready queue which is then executed. This is done continually until all
+ * tasks are complete, or prematurely fails if any of the tasks fail.
+ */
+export class OperationExecutionManager {
+  private readonly _executionRecords: Set<OperationExecutionRecord>;
+  private readonly _runnerGroupRecords: ReadonlyMap<string, Set<OperationExecutionRecord>>;
+  private readonly _runnerGroupRemainingRecords: Map<string, Set<OperationExecutionRecord>>;
+  private readonly _runnerGroupStartTimes: Map<string, number> = new Map();
+  private readonly _parallelism: number;
+  private readonly _totalOperations: number;
+  private readonly _terminal: ITerminal;
+
+  // Variables for current status
+  private _hasReportedFailures: boolean;
+
+  public constructor(operations: Set<Operation>, options: IOperationExecutionManagerOptions) {
+    const { debugMode, parallelism, terminal, loggingManager } = options;
+    this._hasReportedFailures = false;
+    this._terminal = terminal;
+
+    // Convert the developer graph to the mutable execution graph;
+    const executionRecordContext: IOperationExecutionRecordContext = {
+      debugMode,
+      terminal,
+      loggingManager
+    };
+
+    let totalOperations: number = 0;
+    const executionRecords: Map<Operation, OperationExecutionRecord> = new Map();
+    const runnerGroupExecutionRecords: Map<string, Set<OperationExecutionRecord>> = new Map();
+    for (const operation of operations) {
+      const executionRecord: OperationExecutionRecord = new OperationExecutionRecord(
+        operation,
+        executionRecordContext
+      );
+
+      executionRecords.set(operation, executionRecord);
+      const runnerGroupName: string | undefined = executionRecord.runner.groupName;
+      if (runnerGroupName) {
+        const groupExecutionRecords: Set<OperationExecutionRecord> | undefined =
```

Comment:
I think you should just define an `executionRecord.group` property, of type `OperationGroupRecord`:
```ts
class OperationGroupRecord {
  public readonly operations: Set<OperationExecutionRecord> = new Set();
  public readonly remainingOperations: Set<OperationExecutionRecord> = new Set();
  public readonly name: string;

  constructor(name: string) {
    this.name = name;
  }

  public addOperation(operation: OperationExecutionRecord) {
    this.operations.add(operation);
    this.remainingOperations.add(operation);
  }

  public hasFailures(): boolean {
    for (const operation of this.operations) {
      if (operation.status === OperationStatus.Failed) {
        return true;
      }
    }
    return false;
  }
}
```
Also with startTime data (or even a full Stopwatch instance)